### PR TITLE
Allow numba functions within zip files to be cached

### PR DESCRIPTION
**Pull request by max-sixty**
Numba doesn't currently allow `.zip` archives to contain numba functions when caching is enabled. This is all but required when using Spark.

This change enables that, by recognizing a zip file and using a user-wide cache directory for the cache. It closes (most of?) https://github.com/numba/numba/issues/4908#issuecomment-1161476166

I'm the maintainer of numbagg, but haven't contributed directly to numba, so please let me know if the contribution needs any work, happy to make changes.

FWIW — hoping to be helpful — I found it not _super_ easy to use the repo — I notice it doesn't use pytest / code formatters / etc. That does make contributions a bit harder for newcomers, though I'm sure it's for good reason...

### CODE REVIEW

**Comment by gmarkall** [comment]
Thanks for the PR! I'm just marking it as "2 - In progress" while we wait for CI.

> FWIW — hoping to be helpful — I found it not super easy to use the repo — I notice it doesn't use pytest / code formatters / etc. That does make contributions a bit harder for newcomers, though I'm sure it's for good reason...

Thanks for the comments, I do think they're helpful. It doesn't use pytest because there's a lot of homegrown addons to `unittest` that have been added over the years, and porting them over to pytest wasn't trivial (there was an attempt some years ago but I believe it stalled).

WRT to code formatting, we only check for flake8 compliance (with a large number of files excluded as they were historically never checked and they never got fixed up yet). I think there's also more that could be done to make things a smoother experience here, but it's been difficult to move forward with relative to other items.

**Comment by gmarkall** [comment]
From a quick skim it looks like the tests aren't platform-agnostic WRT the way paths are formed (the root, and separators) - I haven't looked deeper, but perhaps `pathlib.Path` could be used to make the form of the tests platform-independent?

**Comment by max-sixty** [comment]
> From a quick skim it looks like the tests aren't platform-agnostic WRT the way paths are formed (the root, and separators) - I haven't looked deeper, but perhaps `pathlib.Path` could be used to make the form of the tests platform-independent?

Thanks, fixed

**Comment by max-sixty** [comment]
> It doesn't use pytest because there's a lot of homegrown addons to `unittest` that have been added over the years, and porting them over to pytest wasn't trivial (there was an attempt some years ago but I believe it stalled).

Yes totally re homegrown addons...

I would say that pytest is much much nicer, both from the perspective of maintainers being able to design better test suites, and the perspective of contributors who are running tests — for example I couldn't seem to enter pdb on a test failure with the current setup? Transitioning probably isn't trivial but also probably not herculean — pytest can collect & run standard unittests, and running two test commands for a while could be a way to get moving.

> WRT to code formatting, we only check for flake8 compliance (with a large number of files excluded as they were historically never checked and they never got fixed up yet). I think there's also more that could be done to make things a smoother experience here, but it's been difficult to move forward with relative to other items.

OK, you might want to check out `ruff` for linting; it fixes most failures itself. For formatting: we moved to `black` in a big bang in xarray, was very successful. The one-off merge conflicts are a bit annoying, but tractable ("merge up to the formatting commit, format with `black`, merge in the formatting commit").

---

The pytest transition probably requires some expertise but if you wanted a hand with linting & formatting, happy to help.

**Comment by max-sixty** [comment]
This is ready for review!

**Comment by gmarkall** [comment]
> This is ready for review!

Oops, sorry, I forgot to check back when I was waiting for CI the other day.

**Comment by max-sixty** [comment]
Hi team — the gentlest of pings here. I know things can get backlogged. I do think this is a fairly small unambiguous improvement, with some decent comments and a great test. If there's anything I can do to reduce the reviewing burden, please lmk.

**Comment by guilhermeleobas** [comment]
Sorry! I'll review it today

**Comment by guilhermeleobas** [comment]
Sorry, I wrongly pushed to your branch an empty commit

**Comment by max-sixty** [comment]
No prob re empty commit, push away.

Docs error seems unrelated though?

```
numba/_helperlib.c:22:44: error: ‘I’ undeclared (first use in this function)
   22 |     #define _complex_float_ctor(r, i) (r + I * i)
      |                                            ^
numba/_helperlib.c:147:12: note: in expansion of macro ‘_complex_float_ctor’
  147 |     *out = _complex_float_ctor((float) _out.real, (float) _out.imag);
      |            ^~~~~~~~~~~~~~~~~~~
numba/_helperlib.c:22:44: note: each undeclared identifier is reported only once for each function it appears in
   22 |     #define _complex_float_ctor(r, i) (r + I * i)
      |                                            ^
numba/_helperlib.c:147:12: note: in expansion of macro ‘_complex_float_ctor’
  147 |     *out = _complex_float_ctor((float) _out.real, (float) _out.imag);
      |            ^~~~~~~~~~~~~~~~~~~
error: command '/usr/bin/gcc' failed with exit code 1
```

**Comment by guilhermeleobas** [comment]
> Docs error seems unrelated though?

Yes! I guess something must have changed on conda-forge.

**Comment by stuartarchibald** [comment]
I think that fixing the docs issue requires merging `numba:main` into this branch so that #9675 is picked up.

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:708]: Move the zip tests to its own class. 

```python
class TestCacheZip(DispatcherCacheUsecaseTest):
    def test_zip_caching(self):
        ...
```

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:730]: Use `setUp/tearDown`

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:741]: Numba has a method to check cache hits/misses:

```python
self.check_hits(fn, hits, misses)
```

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:744]: Maybe you can use `self.import_module()` here. As it will re-import the module.

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:758]: Use `self.check_hits(...)` here.

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:760]: Use `tearDown`

File: numba/core/caching.py
Comment (guilhermeleobas) [inline numba/core/caching.py:346]: This `if` statement will be always true, no? `from_function` will guarantee the `_ZipCacheLocator` will only be created when a zip file is given. On the ctor, `self._zip_path` will never be `None` because there will be a zip subpath.

File: numba/core/caching.py
Comment (guilhermeleobas) [inline numba/core/caching.py:338]: This should never be executed?

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]

File: numba/tests/test_caching.py
Comment (max-sixty) [reply numba/tests/test_caching.py]: I tried but because of the checks that `self.import_module()` does, it fails for the zip. I added a note for future travellers...

File: numba/core/caching.py
Comment (max-sixty) [reply numba/core/caching.py:338]: Yes, agree, thanks for reviewing so thoroughly. The code is now simplified.

File: numba/core/caching.py
Comment (max-sixty) [reply numba/core/caching.py:346]: Yes, agree, thanks for reviewing so thoroughly. The code is now simplified.

**Comment by max-sixty** [comment]
Thanks a lot for the thorough review @guilhermeleobas. I've made those changes and it should be ready for merging / another review.

**Comment by guilhermeleobas** [comment]
closed/reopened your PR to run CI again

**Comment by max-sixty** [comment]
I fixed some flake8 issues. On the topic of dev friendliness:
- `pre-commit` uses a different version of flake8 which doesn't work (at least on my system)
- `flake8` generates lots of errors, because the command that runs in CI isn't just `flake8`

I managed to work this out in a few mins but less easy than other repos I contribute to! (hope this is helpful)

**Comment by esc** [comment]
This is a bit off-topic, but anyway:

I understand that the lack of `pytest` use and a realistic approach to formatting and linting can be quite frustrating as I experience the same. These topics have been discussed repeatedly over the years but there is still no solution in place. I think that is evidence that the tasks are significantly harder than they may appear.

From memory, `pytest` can't be used with Numba as it messes with the bytecode to such an extent that tests start to become unreliable. Numba uses the bytecode as the input to the compiler pipeline, so messing with that was a no-go, not sure if that is still the case. Basically someone would need to take this one, investigate if `pytest` could be used and port the code these days. No small feat.

As for the formatting, any PRs that auto-format are usually quite large and contain many false positives, so the usually just go through multiple rounds of reviews and eventually just become stale and are closed -- as this is not the most exciting task and contributors tend to disengage. Also, auto-formatting (and spelling corrections) will often break stuff, so it needs to be done very, very carefully and each lines needs to be reviewed in earnest. Also, any large scale reformatting PRs tend to break in-flight PRs so that causes additional strain.

If someone were really inclined to help out on this, I think I would be realistically willing to accept about one PR on this topic per week with no more than a handful of changes (maybe 20-30 LOC?) on a single file. This would need to be reviewed very carefully and extensively tested.

One would have to take the files from the `flake8` exclude list:

https://github.com/numba/numba/blob/main/.flake8

Currently that's ~ 230 files or so and so I would estimate the time taken to complete the task -- at a rate of one file per week -- of making the rest of the codebase `flake8` complaint will be about 4 to 5 years (with holidays, release priorities, other random incidents...) -- so perhaps "herculean" is actually the correct adjective here 😇 ? I'm not saying it's impossible, just very tedious and laborious and boring...

Since this is quite off-topic, perhaps it makes sense to move this discussion to a new issue (or a discourse discussion) or find an existing issue where this has been discussed?

**Comment by esc** [comment]
> I fixed some flake8 issues. On the topic of dev friendliness:
> 
> * `pre-commit` uses a different version of flake8 which doesn't work (at least on my system)
> * `flake8` generates lots of errors, because the command that runs in CI isn't just `flake8`
> 
> I managed to work this out in a few mins but less easy than other repos I contribute to! (hope this is helpful)

Here, this could help:

https://github.com/numba/numba/pull/9701

But, this gives new flake8 errors:

```
 💣 zsh» pre-commit run --verbose --all-files
[INFO] Initializing environment for https://github.com/PyCQA/flake8.
[INFO] Installing environment for https://github.com/PyCQA/flake8.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
flake8...................................................................Failed
- hook id: flake8
- duration: 3.41s
- exit code: 1

docs/dagmap/render.py:12:14: E128 continuation line under-indented for visual indent
docs/dagmap/render.py:18:1: W293 blank line contains whitespace
docs/dagmap/render.py:35:9: E128 continuation line under-indented for visual indent
docs/dagmap/render.py:37:81: E501 line too long (84 > 80 characters)
docs/dagmap/render.py:39:81: E501 line too long (110 > 80 characters)
docs/dagmap/render.py:62:81: E501 line too long (84 > 80 characters)
docs/dagmap/render.py:63:81: E501 line too long (100 > 80 characters)
docs/dagmap/render.py:73:81: E501 line too long (82 > 80 characters)
docs/source/developer/compiler_pass_example.py:78:1: E305 expected 2 blank lines after class or function definition, found 1
docs/source/developer/inline_example.py:2:1: F401 'numba' imported but unused
docs/source/developer/inline_overload_example.py:1:1: F401 'numba' imported but unused
docs/source/extending/template.py:6:35: E999 SyntaxError: invalid syntax
docs/source/proposals/np-where-override.py:6:1: E302 expected 2 blank lines, found 1
maint/towncrier_checker.py:49:26: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:6:7: E225 missing whitespace around operator
maint/towncrier_rst_validator.py:11:5: E225 missing whitespace around operator
maint/towncrier_rst_validator.py:30:1: E302 expected 2 blank lines, found 1
maint/towncrier_rst_validator.py:33:36: E225 missing whitespace around operator
maint/towncrier_rst_validator.py:33:65: W291 trailing whitespace
maint/towncrier_rst_validator.py:42:58: W291 trailing whitespace
maint/towncrier_rst_validator.py:43:21: E128 continuation line under-indented for visual indent
maint/towncrier_rst_validator.py:49:57: W291 trailing whitespace
maint/towncrier_rst_validator.py:50:81: E501 line too long (85 > 80 characters)
maint/towncrier_rst_validator.py:51:81: E501 line too long (89 > 80 characters)
maint/towncrier_rst_validator.py:57:1: E305 expected 2 blank lines after class or function definition, found 1
maint/towncrier_rst_validator.py:59:7: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:68:54: W291 trailing whitespace
maint/towncrier_rst_validator.py:84:7: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:86:7: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:97:81: E501 line too long (85 > 80 characters)
maint/towncrier_rst_validator.py:98:81: E501 line too long (91 > 80 characters)
maint/towncrier_rst_validator.py:104:11: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:106:7: F541 f-string is missing placeholders
maint/towncrier_rst_validator.py:110:7: F541 f-string is missing placeholders

pre-commit run --verbose --all-files  15.72s user 1.52s system 187% cpu 9.172 total
```

**Comment by gmarkall** [comment]
> But, this gives new flake8 errors:

I think CI runs `flake8 numba` so this wouldn't usually be an issue - all the errors are from files outside the `numba` directory.

**Comment by max-sixty** [comment]
> > But, this gives new flake8 errors:
> 
> 
> 
> I think CI runs `flake8 numba` so this wouldn't usually be an issue - all the errors are from files outside the `numba` directory.

Configs should generally be in config files rather than in CI scripts, so running `flake8` alone works... I'm guessing there is a `flake8` config that allows specifying which files to check...

**Comment by max-sixty** [comment]
> As for the formatting, any PRs that auto-format are usually quite large and contain many false positives, so the usually just go through multiple rounds of reviews and eventually just become stale and are closed -- as this is not the most exciting task and contributors tend to disengage. Also, auto-formatting (and spelling corrections) will often break stuff, so it needs to be done very, very carefully and each lines needs to be reviewed in earnest. Also, any large scale reformatting PRs tend to break in-flight PRs so that causes additional strain.

To clarify, is this a claim that running `black` over the code would change its functionality?

That would be very surprising to me, and would be a bug in `black`.

Assuming it didn't change its functionality, would you support running black over all the code?

**Comment by max-sixty** [comment]
(I moved the dev tools discussions to other issues. This PR is ready for a second review — thanks!)

File: numba/tests/test_caching.py
Comment (guilhermeleobas) [inline numba/tests/test_caching.py:753]: Isn't this equivalent to call `self.check_hits(test_module.add, 0, 1)`?

**Review by guilhermeleobas — APPROVED** [review]
Code looks good to me. Just wait for @sklam review and we can merge this one. Thanks for your contribution.

**Comment by guilhermeleobas** [comment]
@sklam, are you happy with the current state or would like to review this one as well? If not, we can merge it.

**Comment by sklam** [comment]
@max-sixty, can you add a news fragment for this PR? It will then be part of the release note. See https://github.com/numba/numba/tree/main/docs/upcoming_changes for README and examples.

**Comment by max-sixty** [comment]
Great, added!

File: docs/upcoming_changes/9630.improvement.rst
Comment (esc) [inline docs/upcoming_changes/9630.improvement.rst:1]:
```suggestion
Allow caching of Numba functions within Zip files
```

File: docs/upcoming_changes/9630.improvement.rst
Comment (esc) [inline docs/upcoming_changes/9630.improvement.rst:4]:
```suggestion
This change enables Numba functions imported from a file within a Zip archive to
```

File: docs/upcoming_changes/9630.improvement.rst
Comment (esc) [inline docs/upcoming_changes/9630.improvement.rst:5]:
```suggestion
be cached, by recognizing a Zip file and using a user-wide cache directory
```

File: docs/upcoming_changes/9630.improvement.rst
Comment (esc) [inline docs/upcoming_changes/9630.improvement.rst:6]:
```suggestion
for the cache. Previously, Numba would fail.
```

File: docs/upcoming_changes/9630.improvement.rst
Comment (esc) [inline docs/upcoming_changes/9630.improvement.rst:9]:
```suggestion
Python packages, and heavily used in PySpark.
```

**Review by esc — CHANGES_REQUESTED** [review]
Just some capitalization fixes.

**Comment by max-sixty** [comment]
Done!

**Comment by max-sixty** [comment]
No urgency but it states still Waiting on Author — I think it's now ready for you to merge — lmk if that's not correct

**Review by guilhermeleobas — APPROVED** [review]

**Comment by max-sixty** [comment]
Anything else needed guys?

**Comment by esc** [comment]
> Anything else needed guys?

No, it's fine, just some delays here due to Anaconda-internal CI upgrades taking longer than expected. Hopefully we'll be able to merge this in the next few weeks. I know it's suboptimal, but please be patient, it will get merged eventually.

**Comment by max-sixty** [comment]
Thank you @esc !

**Comment by esc** [comment]
> Thank you @esc !

Apologies for the delay. @sklam made some progress on the recent 3.12 issue our test-suite was hitting:

https://github.com/numba/numba/pull/9727

so with a little luck this PR will be merged early next week.

**Comment by esc** [comment]
thank you for the patch

**Pull request merged**
